### DISCUSSION TITLE
Add NavSatFix as Displayable topic in 3d Panel

### DISCUSSION THREAD

**Discussion by drldcsta** [foxglove/community#319] [Q&A]
**Current Options**

Currently, the map is only available for display on the map panel, which makes it difficult to tell where exactly the robot is.

**Proposal**
We propose having the NavSatFix as a displayable topic in the 3d panel, and this would render map tiles when toggled on. In addition to this, we think it would be beneficial to switch to Mapbox for map tiles, as opposed to the current OpenStreetMap solution

**Comment by defunctzombie** [comment]
To display a message in the 3d panel, we need to know the `frame` of the message and also have a [FrameTransform](https://foxglove.dev/docs/studio/messages/frame-transform) between the `frame` and whatever other objects (and their frames) are in the same scene.

How would we go about doing that for a NavSatFix or [LocationFix](https://foxglove.dev/docs/studio/messages/location-fix) message?

**Comment by AndrewNolte** [comment]
> To display a message in the 3d panel, we need to know the `frame` of the message and also have a [FrameTransform](https://foxglove.dev/docs/studio/messages/frame-transform) between the `frame` and whatever other objects (and their frames) are in the same scene.
> 
> How would we go about doing that for a NavSatFix or [LocationFix](https://foxglove.dev/docs/studio/messages/location-fix) message?

[NavSatFix](https://docs.ros.org/en/api/sensor_msgs/html/msg/NavSatFix.html) has a header, which has a frame_id, so this could pin down the location. What could be tricky here is aligning the heading/orientation correctly, which isn't in this message. Maybe this could be encoded somehow in the global tf that `frame_id` has.

**Comment by defunctzombie** [comment]
> [NavSatFix](https://docs.ros.org/en/api/sensor_msgs/html/msg/NavSatFix.html) has a header, which has a frame_id, so this could pin down the location.

The transforms have information about how one frame relates to another (usually in units like meters and radians). How would you then tie that to the lat/long of the NavSatFix?

The issue I'm stuck on:

I've got a 3d panel visualizing in `base_link`. I receive a NavSatFix message with frame_id `gps_sensor`. If I have a transform from `gps_sensor` to `base_link`, that tells me the location of `gps_sensor` in `base_link`. How would I figure out the location of lat/long from that message in terms of (meters and radians) relative to `gps_sensor` tho?

**Comment by OgreTransporter** [comment]
Mapbox I don't find absolutely necessary, but I agree that there should be alternative map sources. The custom layer does not work properly, unfortunately. Also, an offline alternative (see foxglove/community#273) would be useful.

**Comment by cshang412** [comment]
i suggest visual like rviz satellite tools, there is already a  tool for display the satellite map in Foxglove window,  i think intergrate this feature to 3D panel's custom layer is an  better choice .
![图片](https://github.com/foxglove/community/assets/29396384/c251b4dc-5ac9-40b0-b782-5d800532fbac)
![图片](https://github.com/foxglove/community/assets/29396384/15930697-d4d4-48c9-9cf8-dbe0c3b28634)
